documentation/FIR-MLIR-dialect.md

```diff
@@ -0,0 +1,204 @@
+
+### What is MLIR?
+
+The MLIR project is a recently open-sourced project on Github. MLIR (see [https://github.com/tensorflow/mlir](https://github.com/tensorflow/mlir) for more information on MLIR) is being built as a common multi-level intermediate representation (IR) for the TensorFlow machine learning tools. It provides a common IR framework for building multiple layers of representation called dialects. These dialects can be transformed (converted) from one to another allowing for optimizations and successive lowering of the IR down to backend code generators. One of these code generators is LLVM IR. MLIR is explicitly intended to be extensible such that other projects can add their own dialects.
+
+  
+
+Out of the box, MLIR supports a refined polyhedral model for transforming affine operations on array data. These mathematically-based representations will help realize F18's higher level goals of composing, transforming, and optimizing Fortran array operations for HPC.
+
+### What is meant by a FIR dialect?
+
+MLIR has no support for Fortran out of the box. Support for Fortran must be built and added through the extensibility features of MLIR. This process of extending MLIR is called adding a dialect. In our case, we'll add a FIR dialect for representation of Fortran programs.
+
+Specifically, this means that we can proceed with the plan as laid out in the Fortran IR design document with the advantages of not having to build from scratch our own
+
+-   IR infrastructure
+    
+-   LLVM IR bridge
+    
+-   Common optimization passes (For example: CSE, DCE, loop fusion, loop invariant code motion, loop tiling, loop unroll and jam, loop unrolling, vectorization, etc.)
+    
+
+A dialect can be thought of as its own distinct IR, where the design and semantics of the IR are domain-specific for the purposes of the dialect. Specifically, MLIR supports a modern IR design with
+
+-   A predefined set of operations with precise semantics
+    
+-   Structural/logical grouping of sets of operations
+    
+-   Explicit control flow
+    
+-   Explicit data-flow
+    
+-   Strong typing
+    
+-   Meta information
+    
+
+This document describes FIR, which is a set of extensions upon the standard MLIR dialect. When lowering a Fortran parse tree to FIR, the compiler will produce a mix of operations defined in the FIR dialect, the MLIR standard dialect, and other pre-defined dialects, such as the affine dialect.
```

Comment:
Please add a reference for the affine dialect.

Reply:
I'll see if there is one.

Reply:
Done